File: frontend/src/api/books.ts
import { apiFetch, apiFetchForm } from "./client";
import { API_BASE_URL } from "../lib/constants";

export type BookLemmaSample = {
  id: number;
  lemma: string;
  frequency: number;
  rank: number;
  in_study_set: boolean;
  is_hidden: boolean;
  matched_baseline: boolean;
};

export type BookProgress = {
  learner_id: number;
  study_known: number;
  study_total: number;
  study_progress_percent: number;
  page_known: number;
  content_total: number;
  page_coverage_percent: number;
  ready_to_read: boolean;
  days_estimate: number;
  learning_count: number;
  familiar_count: number;
  mastered_count: number;
};

export type BookSummary = {
  id: number;
  title: string;
  title_source?: string;
  title_needs_review?: boolean;
  original_filename: string;
  status: string;
  coverage_target: number;
  token_count: number;
  unique_lemma_count: number;
  content_lemma_count: number;
  study_lemma_count: number;
  skipped_function_words: number;
  skipped_proper_nouns: number;
  coverage_curve: Record<string, number>;
  days_at_five_new: number;
  word_list_id: number | null;
  assigned_learner_ids: number[];
  analysis_engine: string;
  confirmed_at: string | null;
  created_at: string | null;
  baseline_match_count?: number;
  new_word_count?: number;
  sample_study?: BookLemmaSample[];
  sample_advanced?: BookLemmaSample[];
};

export async function previewBook(file: File): Promise<BookSummary> {
  const formData = new FormData();
  formData.append("file", file);
  return apiFetchForm<BookSummary>("/books/preview", formData, API_BASE_URL);
}

export async function listBooks(): Promise<BookSummary[]> {
  const response = await apiFetch<{ books: BookSummary[] }>("/books", {}, API_BASE_URL);
  return response.books;
}

export async function getBook(bookId: number): Promise<BookSummary> {
  return apiFetch<BookSummary>(`/books/${bookId}`, {}, API_BASE_URL);
}

export async function confirmBook(
  bookId: number,
  coverageTarget?: number,
  title?: string,
): Promise<BookSummary> {
  return apiFetch<BookSummary>(
    `/books/${bookId}/confirm`,
    {
      method: "POST",
      body: JSON.stringify({ coverage_target: coverageTarget ?? null, title: title ?? null }),
    },
    API_BASE_URL,
  );
}

export async function updateBookTitle(bookId: number, title: string): Promise<BookSummary> {
  return apiFetch<BookSummary>(
    `/books/${bookId}`,
    { method: "PATCH", body: JSON.stringify({ title }) },
    API_BASE_URL,
  );
}

export async function assignBook(bookId: number, learnerId: number): Promise<BookSummary> {
  return apiFetch<BookSummary>(
    `/books/${bookId}/assign`,
    { method: "POST", body: JSON.stringify({ learner_id: learnerId }) },
    API_BASE_URL,
  );
}

export async function unassignBook(bookId: number, learnerId: number): Promise<BookSummary> {
  return apiFetch<BookSummary>(`/books/${bookId}/assign/${learnerId}`, { method: "DELETE" }, API_BASE_URL);
}

export async function deleteBook(bookId: number): Promise<void> {
  await apiFetch<void>(`/books/${bookId}`, { method: "DELETE" }, API_BASE_URL);
}

export async function hideBookLemma(
  bookId: number,
  lemmaId: number,
  hidden: boolean,
): Promise<BookSummary> {
  return apiFetch<BookSummary>(
    `/books/${bookId}/lemmas/${lemmaId}`,
    { method: "PATCH", body: JSON.stringify({ hidden }) },
    API_BASE_URL,
  );
}

export type SuspiciousLemma = {
  id: number;
  lemma: string;
  frequency: number;
  rank: number;
  in_study_set: boolean;
  is_hidden: boolean;
  reason: string;
};

export async function getSuspiciousLemmas(
  bookId: number,
  includeHidden = false,
): Promise<SuspiciousLemma[]> {
  const params = includeHidden ? "?include_hidden=true" : "";
  const response = await apiFetch<{ items: SuspiciousLemma[]; total: number }>(
    `/books/${bookId}/suspicious-lemmas${params}`,
    {},
    API_BASE_URL,
  );
  return response.items;
}

export async function bulkHideBookLemmas(
  bookId: number,
  lemmaIds: number[],
  hidden: boolean,
): Promise<BookSummary> {
  return apiFetch<BookSummary>(
    `/books/${bookId}/lemmas/bulk-hide`,
    { method: "POST", body: JSON.stringify({ lemma_ids: lemmaIds, hidden }) },
    API_BASE_URL,
  );
}

export async function getBookProgress(bookId: number, learnerId?: number): Promise<BookProgress[]> {
  const params = learnerId != null ? `?learner_id=${learnerId}` : "";
  return apiFetch<BookProgress[]>(`/books/${bookId}/progress${params}`, {}, API_BASE_URL);
}

export async function getMyBookProgress(): Promise<BookProgress | null> {
  return apiFetch<BookProgress | null>("/loop/book-progress", {}, API_BASE_URL);
}

export type BookDefinitionsSummary = {
  needs_refresh_count: number;
  missing_en_count: number;
  missing_zh_count: number;
};

export type DefinitionFillJob = {
  id: number;
  status: string;
  total: number;
  processed: number;
  filled: number;
  failed: number;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
};

export async function getBooksDefinitionsSummary(): Promise<BookDefinitionsSummary> {
  return apiFetch<BookDefinitionsSummary>("/books/definitions-summary", {}, API_BASE_URL);
}

export async function startBookDefinitionFillJob(): Promise<DefinitionFillJob> {
  return apiFetch<DefinitionFillJob>(
    "/books/fill-definitions",
    { method: "POST" },
    API_BASE_URL,
  );
}

export async function getCurrentBookDefinitionFillJob(): Promise<DefinitionFillJob | null> {
  return apiFetch<DefinitionFillJob | null>(
    "/books/fill-definitions/current",
    {},
    API_BASE_URL,
  );
}

export async function cancelBookDefinitionFillJob(jobId: number): Promise<DefinitionFillJob> {
  return apiFetch<DefinitionFillJob>(
    `/books/fill-definitions/${jobId}/cancel`,
    { method: "POST" },
    API_BASE_URL,
  );
}

export type PlaceholderLemma = {
  id: number;
  book_id: number;
  book_title: string;
  lemma: string;
  frequency: number;
  in_study_set: boolean;
  is_hidden: boolean;
};

export async function getPlaceholderLemmas(
  jobId?: number,
  includeHidden = false,
): Promise<PlaceholderLemma[]> {
  const params = new URLSearchParams();
  if (jobId != null) {
    params.set("job_id", String(jobId));
  }
  if (includeHidden) {
    params.set("include_hidden", "true");
  }
  const query = params.toString();
  const response = await apiFetch<{ items: PlaceholderLemma[]; total: number }>(
    `/books/placeholder-lemmas${query ? `?${query}` : ""}`,
    {},
    API_BASE_URL,
  );
  return response.items;
}

File: backend/app/api/v1/books.py
from fastapi import APIRouter, Depends, File, Query, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import require_parent
from app.core.rate_limit import limiter
from app.database import get_db
from app.models.user import User
from app.schemas.book import (
    BookAssignRequest,
    BookConfirmRequest,
    BookDefinitionsSummary,
    BookLemmaBulkHideRequest,
    BookLemmaHideRequest,
    BookListResponse,
    BookProgress,
    BookSummary,
    BookUpdateRequest,
    DefinitionFillJobResponse,
    PlaceholderLemmaListResponse,
    SuspiciousLemmaListResponse,
)
from app.services import book_service
from app.services.word_bank_service import job_to_dict

router = APIRouter(prefix="/books", tags=["books"])


@router.post("/preview", response_model=BookSummary)
async def preview_book(
    file: UploadFile = File(...),
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    limiter.check(key=f"book-preview:{parent.id}", limit=10, window_seconds=60 * 60)
    book = await book_service.preview_upload(db, parent_id=parent.id, upload=file)
    return BookSummary(**book_service.book_to_preview(book))


@router.get("", response_model=BookListResponse)
async def list_books(
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookListResponse:
    books = await book_service.list_books(db, parent.id)
    summaries: list[BookSummary] = []
    for book in books:
        assigned = await book_service.assigned_learner_ids(db, book)
        summaries.append(
            BookSummary(**book_service.book_to_summary(book, assigned_learner_ids=assigned))
        )
    return BookListResponse(books=summaries)


@router.get("/definitions-summary", response_model=BookDefinitionsSummary)
async def books_definitions_summary(
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookDefinitionsSummary:
    data = await book_service.get_books_definitions_summary(db, parent.id)
    return BookDefinitionsSummary(**data)


@router.post("/fill-definitions", response_model=DefinitionFillJobResponse)
async def start_book_definition_fill(
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> DefinitionFillJobResponse:
    data = await book_service.start_book_definition_fill_job(db, parent.id)
    return DefinitionFillJobResponse(**data)


@router.get("/fill-definitions/current", response_model=DefinitionFillJobResponse | None)
async def current_book_definition_fill(
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> DefinitionFillJobResponse | None:
    job = await book_service.get_current_book_definition_fill_job(db, parent.id)
    if job is None:
        return None
    return DefinitionFillJobResponse(**job_to_dict(job))


@router.post("/fill-definitions/{job_id}/cancel", response_model=DefinitionFillJobResponse)
async def cancel_book_definition_fill(
    job_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> DefinitionFillJobResponse:
    data = await book_service.cancel_book_definition_fill_job(db, parent.id, job_id)
    return DefinitionFillJobResponse(**data)


@router.get("/placeholder-lemmas", response_model=PlaceholderLemmaListResponse)
async def list_placeholder_lemmas(
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
    job_id: int | None = Query(default=None),
    include_hidden: bool = Query(default=False),
) -> PlaceholderLemmaListResponse:
    items = await book_service.list_book_placeholder_lemmas(
        db,
        parent_id=parent.id,
        job_id=job_id,
        include_hidden=include_hidden,
    )
    return PlaceholderLemmaListResponse(items=items, total=len(items))


@router.delete("/{book_id}", status_code=204)
async def delete_book(
    book_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> None:
    await book_service.delete_book(db, parent_id=parent.id, book_id=book_id)


@router.patch("/{book_id}", response_model=BookSummary)
async def update_book(
    book_id: int,
    payload: BookUpdateRequest,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service.update_book_title(
        db, parent_id=parent.id, book_id=book_id, title=payload.title
    )
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_preview(book, assigned_learner_ids=assigned))


@router.get("/{book_id}", response_model=BookSummary)
async def get_book(
    book_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service._get_book_for_parent(db, book_id, parent.id)
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_preview(book, assigned_learner_ids=assigned))


@router.post("/{book_id}/confirm", response_model=BookSummary)
async def confirm_book(
    book_id: int,
    payload: BookConfirmRequest,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service.confirm_book(
        db,
        parent_id=parent.id,
        book_id=book_id,
        coverage_target=payload.coverage_target,
        title=payload.title,
    )
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_preview(book, assigned_learner_ids=assigned))


@router.post("/{book_id}/assign", response_model=BookSummary)
async def assign_book(
    book_id: int,
    payload: BookAssignRequest,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service.activate_for_learner(
        db, parent_id=parent.id, book_id=book_id, learner_id=payload.learner_id
    )
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_summary(book, assigned_learner_ids=assigned))


@router.delete("/{book_id}/assign/{learner_id}", response_model=BookSummary)
async def unassign_book(
    book_id: int,
    learner_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    await book_service.deactivate_for_learner(
        db, parent_id=parent.id, book_id=book_id, learner_id=learner_id
    )
    book = await book_service._get_book_for_parent(db, book_id, parent.id)
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_summary(book, assigned_learner_ids=assigned))


@router.patch("/{book_id}/lemmas/{lemma_id}", response_model=BookSummary)
async def hide_book_lemma(
    book_id: int,
    lemma_id: int,
    payload: BookLemmaHideRequest,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service.hide_lemma(
        db, parent_id=parent.id, book_id=book_id, lemma_id=lemma_id, hidden=payload.hidden
    )
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_preview(book, assigned_learner_ids=assigned))


@router.get("/{book_id}/suspicious-lemmas", response_model=SuspiciousLemmaListResponse)
async def list_suspicious_lemmas(
    book_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
    include_hidden: bool = Query(default=False),
) -> SuspiciousLemmaListResponse:
    items = await book_service.list_suspicious_lemmas(
        db, parent_id=parent.id, book_id=book_id, include_hidden=include_hidden
    )
    return SuspiciousLemmaListResponse(items=items, total=len(items))


@router.post("/{book_id}/lemmas/bulk-hide", response_model=BookSummary)
async def bulk_hide_book_lemmas(
    book_id: int,
    payload: BookLemmaBulkHideRequest,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
) -> BookSummary:
    book = await book_service.bulk_hide_lemmas(
        db,
        parent_id=parent.id,
        book_id=book_id,
        lemma_ids=payload.lemma_ids,
        hidden=payload.hidden,
    )
    assigned = await book_service.assigned_learner_ids(db, book)
    return BookSummary(**book_service.book_to_preview(book, assigned_learner_ids=assigned))


@router.get("/{book_id}/progress", response_model=list[BookProgress])
async def book_progress(
    book_id: int,
    parent: User = Depends(require_parent),
    db: AsyncSession = Depends(get_db),
    learner_id: int | None = Query(default=None),
) -> list[BookProgress]:
    book = await book_service._get_book_for_parent(db, book_id, parent.id)
    if learner_id is not None:
        await book_service._parent_learner(db, parent_id=parent.id, learner_id=learner_id)
        data = await book_service.progress_for_learner(db, book, learner_id)
        return [BookProgress(**data)]
    assigned = await book_service.assigned_learner_ids(db, book)
    results: list[BookProgress] = []
    for assigned_id in assigned:
        data = await book_service.progress_for_learner(db, book, assigned_id)
        results.append(BookProgress(**data))
    return results
